### DISCUSSION TITLE
Proxy mode

### DISCUSSION THREAD

**Discussion by cofob** [yggdrasil-network/yggdrasil-go#850] [Ideas]
Proxy mode (`http`/`socks`) like in `i2p` or `tor`. This will allow you to use full node addresses (for example, `<publickey>.yggip` to protect against address spoofing) and also run `yggdrasil` without superuser rights (now you need to create a `tun` interface).

**Comment by cofob** [comment]
It can also be useful for building a browser extension. (#381)

**Comment by neilalexander** [comment]
Dupe of #745.

**Comment by xialvjun** [comment]
Proxy Mode can also be run in non **`root`** permission.

**Comment by neilalexander** [comment]
This discussion comes up a lot and the reason we don't do it is always the same — in order for proxy mode to be able to talk to any non-proxying Yggdrasil host, we would need to embed most of an IPv6 stack and a TCP implementation into Yggdrasil. That's a lot of code, a lot of surface area to go wrong and more than we really need at this stage for an experimental project.

**Reply by majestrate** [reply]
there is [gvisor's netstack](https://pkg.go.dev/gvisor.dev/gvisor/pkg/sentry/socket/netstack) but that project tends to be highly temperamental as an upstream dep. from my own passive observations and mildly effort exerted toys that use them i wish they were a lot more stable.

**Comment by andyoknen** [comment]
@neilalexander 

Hi! I'm not very good at network programming, so I need some advice.

Allow myself a little introduction. We am working on a decentralized network infrastructure on top of blockchain nodes, in which each blockchain node is a full-fledged web server for the "web 3 ecosystem" - https://github.com/pocketnetteam/get.started

**My main question is how to link all the elements of the ecosystem through yggdrasil without using a virtual interface at the operating system level.**

I understand that it is impossible to link yggdrasil nodes in proxy mode and regular nodes. **But is it possible to implement yggdrasil in each server node and client application for mutual "communication"?**

The main obstacles to using yggdrasil "as is" are the additional installed software and the mandatory firewall configuration for safe use.

At the moment, I am only studying the user side of the software, and this is extremely successful for me. In the near future, I plan to dive deeper into code study and integration possibilities (static compilation in C++ nodes and dynamic library for use in the Node.js client side).

**Reply by neilalexander** [reply]
It is definitely possible to use Yggdrasil in this way but only easily within a Go application. Yggdrasil matches the `net.PacketConn` interface, similar to UDP sockets, so you can embed a node and send packets that way by addressing them to a public key. You can also layer QUIC on top for streaming/reliable/ordered semantics, see https://github.com/yggdrasil-network/yggquic for an example of this.

**Reply by andyoknen** [reply]
That is, for use in other code (c++, nodejs) I need a small library that will wrap net.PacketConn interface? To minimize the work in our software, it will probably be necessary to implement a pseudo HTTP interface. It sounds reasonable and quite feasible.

**Reply by andyoknen** [reply]
I have now thought of another option - to limit traffic through yggdrasil to only certain ports in order to minimize risks without configuring a firewall. Is this possible in the current implementation?

**Comment by zacoons** [comment]
Can this be marked as answered now? It seems like this has been implemented in https://github.com/yggdrasil-network/yggstack
there's also https://github.com/yggdrasil-network/yggquic
